keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,1182 @@
+doesn't support the NodeMaintenance API, kubectl will perform the node drain in a backwards
+compatible way.
+
+`kubectl uncordon` will be amended with `--node-maintenance` option. This can be used to delete the
+NodeMaintenance object together with marking the node as schedulable again. The default value of
+this flag will be `false`.
+
+### NodeMaintenance API
+
+NodeMaintenance objects serve as an intent to remove or migrate pods from a particular node. We will
+also support setting a reason for the maintenance.
+
+```golang
+
+type NodeMaintenance struct {
+    ...
+    Spec NodeMaintenanceSpec
+    Status NodeMaintenanceStatus
+}
+
+type NodeMaintenanceSpec struct {
+    NodeName string
+    Reason string
+}
+
+type NodeMaintenanceStatus struct {
+    // number of pods this node maintenance is requesting to terminate
+    TargetedPods int32
+}
+```
+
+The NodeMaintenance API is being introduced with a minimal set of features, but it is expected that
+additional features will be added to the API in the future.
+
+The design of the API should not assume that only one node is selected for the node maintenance.
+And a Node to NodeMaintenance mapping should not be used or encouraged. `.spec.nodeName` field
+could be made optional if the requirements change in the future, in favor of `.spec.nodeSelector`.
+This could cause pods on multiple nodes to be marked for the maintenance at the same time, so even
+the NodeMaintenance `.status` should be made general enough.
+
+<<[UNRESOLVED NodeMaintenance ]>>
+
+- It is unclear whether we should allow marking the NodeMaintenances as finished and keeping the
+  old maintenances in the system for bookkeeping purposes.
+
+<<[/UNRESOLVED]>>
+
+### NodeMaintenance Controller
+
+Node maintenance controller will be introduced and added to `kube-controller-manager`. It will
+observe NodeMaintenance objects and set `TargetedByNodeMaintenance` condition on selected pods.
+The pods would be selected according to the node. Controllers that own these pods, would observe
+them and start a removal or migration from these nodes upon detecting this condition.
+
+The selection process can be later enhanced to include priority or pod type (to target daemon sets
+which we do not currently support).
+
+The controller would also remove the `TargetedByNodeMaintenance` condition from the targeted pods
+upon premature `NodeMaintenance` object removal.
+
+The controller can then show progress by reconciling `.status.targetedPods`, to show how many pods
+remain to be removed from the node.
+
+The controller would also emit a warning event if the targeted node was not marked as unschedulable.
+We could prevent this, by stomping on any changes to the node's `.spec.unschedulable` field and set
+it to true for the whole lifetime of `NodeMaintenance` object.
+
+<<[UNRESOLVED NodeMaintenance .status ]>>
+
+- Should we target all pods or just the ones we know can be deleted (drain filters)? The later will
```

Comment:
> Ideally, we'd target all pods that we'd like to be removed.

knowing which pods we (i.e. the admin) would like to be removed is the difficult part. Since we would like to automate the more difficult cases (drain filters) it would be hard to do drastic actions in the future if you cannot opt out from them. That is, if we mark them now, and removal feature is implemented in the controllers later, we could break users.

To support the observability (manual actions) we could just count them in the status instead for now (e.g. `NonTargetedPods`).

Reply:
I have tried to include most of the drain filters and add a reasoning why we can include them. Nevertheless I think we still have to skip DaemonSet pods as they can have critical services and should be opted in.

Would be great to get feedback about the pods with local storage though.